src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/servlets/SearchServlet.java

```diff
@@ -15,24 +15,100 @@
 package com.google.step.YOUR_PROJECT_NAME_HERE.servlets;
 
 import java.io.IOException;
-import java.util.Arrays;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.json.JSONObject;
 
 /** Servlet that handles searches. */
 @WebServlet("/search")
 public class SearchServlet extends HttpServlet {
 
+  private static final String W3_CSE_ID = "010448377421452380243:pat2rmwjvb8";
+  private static final String STACK_CSE_ID = "010448377421452380243:6b8rs1ze5oy";
+  private static final String GEEKS_CSE_ID = "010448377421452380243:ieq7z84z2qq";
+  private static final String API_KEY = "AIzaSyCMg08fxt9IX8LOAdwJGR0DyphMFpXPe5k";
+  private static final String CSE_URL =
+      "https://www.googleapis.com/customsearch/v1?key=" + API_KEY + "&cx=";
+
+  private final HttpClient httpClient =
+      HttpClient.newBuilder().version(HttpClient.Version.HTTP_2).build();
+
+  private static String encodeValue(String value) {
+    try {
+      return URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
-    // Set some data on the request for the template to access.
-    request.setAttribute("previousQuery", request.getParameter("q"));
-    request.setAttribute("results", Arrays.asList("Aaron", "Iris", "Kayla"));
-    // Forward the request to the template (which is a servlet itself).
-    request.getRequestDispatcher("WEB-INF/templates/search.jsp").forward(request, response);
+    String query = "&q=" + encodeValue(request.getParameter("q"));
+    List<String> allLinks = new ArrayList<>();
+
+    // TODO: after implementing scraping, consider changing try/catch calls to a for-loop
+    /** Send request to retrieve card content through w3School site link from Google CSE */
+    try {
+      String w3Link = getLink(CSE_URL + W3_CSE_ID + query);
```

Comment:
There's a lot of URI fragments being passed around.

CSE_URL is not a valid anything by itself since it ends with &cx=.
W3_CSE_ID is an ID by itself which is fine.
query is not a valid anything since it starts with &q=.

I think it's easier to remember these restrictions if each variable is a valid _something_ - e.g. CSE_URL is just a URL as advertised without the &cx=, and query is just the query as advertised without the &q=, and you insert these connectors when you build the URI.

Note that the URI building is split into two pieces - it's done here, but finished within getLink at URI.create. How about just creating the whole URI in one of the two places?